src/odemis/driver/keysight.py

```diff
@@ -0,0 +1,333 @@
+            self.open()
+
+    def open(self, ip_addr=None, tcp_timeout=None):
+        """Open/Reopen Connection
+
+        If a given argument is None, the previous value of the same argument is used.
+
+        :param ip_addr: the instrument's IP-Address (digits and dots format)
+        :param tcp_timeout: TCP-Socket time-out (in seconds)
+        """
+        self.close()
+
+        if ip_addr is not None:
+            self._ip_addr = ip_addr
+        if tcp_timeout is not None:
+            self._tcp_timeout = float(tcp_timeout)
+
+        # Open TCP-IP Socket:
+        self._tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.IPPROTO_TCP)
+        self._tcp_sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        self._tcp_sock.settimeout(self._tcp_timeout)
+        self._tcp_sock.connect((self._ip_addr, self._tcp_port))
+
+    def close(self):
+        # Close Connection
+        if self._tcp_sock is not None:  # pass if not connectedS
+            self._tcp_sock.close()
+            self._tcp_sock = None
+
+    def sendQuery(self, query_str):
+        # Send the given query to the instrument and read the response.
+        if self._tcp_sock is not None:  # pass if not connected
+
+            query_str += '\r\n'
+            query_byte = query_str.encode()
+            self._tcp_sock.sendall(query_byte)
+            resp = self.read_resp()
+            return resp
+
+    def readResp(self):
+        # Read response from the instrument.
+        if self._tcp_sock is not None:  # pass if not connected
+            resp = []
+            ch = self._tcp_sock.recv(1).decode()
+            while "\n" != ch:
+                if "\r" != ch:
+                    resp.append(ch)
+                ch = self._tcp_sock.recv(1).decode()
+            return ''.join(resp)
+
+    def sendCmd(self, cmd_str, wait_for_complete=True):
+        # Send the given command to the instrument.
+        if self._tcp_sock is not None:  # pass if not connected
+            cmd_str = cmd_str.rstrip()
+            cmd_str += '\n'
+            cmd_byte = cmd_str.encode()
+            self._tcp_sock.sendall(cmd_byte)  # send command
+
+
+class Keysight33622ASimulator(object):
+    """
+    Simulates a Keysight 33622A waveform generator
+    """
+
+    def __init__(self, timeout=1):
+        self.timeout = timeout
+        self._output_buf = ""  # what the commands sends back to the "host computer"
+        self._input_buf = ""  # what we receive from the "host computer"
+
+        self._output_states = {
+            'channel1': {
+                'tracking': 'on',  # on, off, invert
+                'voltage_low': -5.0,  # V
+                'voltage_high': 5.0,  # V
+                'voltage_limit_low': -5.0,  # V
+                'voltage_limit_high': 5.0,  # V
+                'period': 150,  # 1/frequency in Hz (1 μHz to 50 MHz)
+                'duty_cycle': 50  # percentage
+            },
+            'channel2': {
```

Comment:
As all the channels have the same attributes, I think it would make more sense to create a dedicated class (or use namedtuple, if you like it better)